File: java/src/main/java/leetcode_problem/SubArrayProduct.java
package leetcode_problem;

import java.util.*;

public class SubArrayProduct {

    // Brute-force approach to count subarrays with product less than k
    public static int maxProductArrayBruteForce(int[] nums, int k) {
        int count = 0;

        for (int i = 0; i < nums.length; i++) {
            // Skip elements greater than or equal to k
            if (nums[i] >= k) {
                continue;
            }

            int product = 1;
            for (int j = i; j < nums.length; j++) {
                product *= nums[j];

                // If the product exceeds k, stop further processing
                if (product >= k) {
                    break;
                }
                count++;
            }
        }
        return count;
    }

    // Sliding window approach to count subarrays with product less than k
    public static int subArrayProductSlidingWindow(int[] nums, int k) {
        if (k <= 1) return 0;

        int i = 0, j = 0, product = 1, count = 0;
        int n = nums.length;

        while (j < n) {
            product *= nums[j];

            // Shrink the window while product is greater than or equal to k
            while (product >= k && i <= j) {
                product /= nums[i];
                i++;
            }

            // Number of valid subarrays ending at j
            count += j - i + 1;
            j++;
        }
        return count;
    }

    // Logarithmic approach using binary search to count subarrays with product less than k
    public static int subArrayLogarithmic(int[] nums, int k) {
        if (k == 0) return 0;

        double logK = Math.log(k);
        List<Double> logsSum = new ArrayList<>();
        logsSum.add(0.0);

        // Compute cumulative logarithmic sum
        for (int num : nums) {
            logsSum.add(logsSum.get(logsSum.size() - 1) + Math.log(num));
        }

        int count = 0;
        int n = logsSum.size();

        for (int i = 0; i < n; i++) {
            int low = i + 1, high = n;

            // Binary search for the rightmost valid subarray
            while (low < high) {
                int mid = (low + high) / 2;

                if (logsSum.get(mid) < logsSum.get(i) + logK - 1e-9) {
                    low = mid + 1;
                } else {
                    high = mid;
                }
            }

            count += low - i - 1;
        }
        return count;
    }

    public static void main(String[] args) {
        int[] nums = {10, 5, 2, 6};
        int k = 100;

        System.out.println("Brute Force: " + maxProductArrayBruteForce(nums, k));
        System.out.println("Sliding Window: " + subArrayProductSlidingWindow(nums, k));
        System.out.println("Logarithmic Approach: " + subArrayLogarithmic(nums, k));
    }
}

File: python/leetcode_problems/subarray_product.py
import math

def sub_array_logarithmic(nums, k):
    # Handle edge case where k is 0
    if k == 0:
        return 0

    # Precompute the logarithm of k for efficiency
    logK = math.log(k)

    # Initialize a list to store the cumulative sum of logarithms of nums
    logs_sum = [0]

    # Calculate the cumulative sum of logarithms
    for num in nums:
        logs_sum.append(logs_sum[-1] + math.log(num))

    # Initialize count of sub arrays with product less than k
    count = 0

    # Iterate over each possible start of a subarray
    for i in range(len(logs_sum)):
        # Initialize binary search range for end of subarray
        low, high = i + 1, len(logs_sum)

        # Perform binary search to find the end of the subarray
        while low < high:
            mid = (low + high) // 2

            # Check if the product of the subarray is less than k
            # A small tolerance (-1e-9) is used to handle floating-point precision issues.
            if logs_sum[mid] < logs_sum[i] + logK - 1e-9:
                # If true, move the low pointer to mid + 1
                low = mid + 1
            else:
                # If false, move the high pointer to mid
                high = mid

        # Increment count by the number of valid sub arrays ending at low - 1
        count += low - i - 1

    # Return the total count of sub arrays with product less than k
    return count
def max_product_array_brute_force(num,k):

    count = 0

    for i in range(len(nums)):

        if nums[i] >= k:
            continue
        curr = 1

        for j in range(i,len(nums)):
            curr = nums[j] * curr

            if curr >= k:
                break
            count += 1


    return count

def sub_array_product_sliding_window(nums,k):

    if k <= 1:
        return 0

    i , j, product = 0,0,1

    n = len(nums)

    count = 0


    while j < n:
        product = product * nums[j]

        while product >= k:
            product = product // nums[i]
            i += 1

        count += j - i + 1
        j += 1


    return count


if __name__ == '__main__':

    nums = [10, 5, 2, 6]
    k = 100
    print(max_product_array_brute_force(nums,k))
    print(sub_array_product_sliding_window(nums,k))
    print(sub_array_logarithmic(nums,k))









